cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/SecureMultiStore.java

```diff
@@ -0,0 +1,267 @@
+  private final SecureStoreManager secureStoreManager;
+
+  public SecureMultiStore(SecureStore secureStore, SecureStoreManager secureStoreManager,
+      TransactionRunner transactionRunner) {
+    this.secureStore = secureStore;
+    this.secureStoreManager = secureStoreManager;
+    this.transactionRunner = transactionRunner;
+  }
+
+  /**
+   * An identifier that uniquely identifies the class which utilizes secure store.
+   */
+  abstract String getResourceType();
+
+  abstract R constructFromSecureData(T metadata, String secureData);
+
+  abstract Optional<VersionedObject<T>> getMetadataInternal(StructuredTableContext context,
+      NamespaceResourceReference ref) throws IOException;
+
+  abstract Collection<NamespaceResourceReference> listMetadataInternal(
+      StructuredTableContext context, String namespace) throws IOException;
+
+  abstract void writeMetadataInternal(StructuredTableContext context,
+      NamespaceResourceReference ref, T obj, String version) throws IOException;
+
+  abstract void deleteMetadataInternal(StructuredTableContext context,
+      NamespaceResourceReference ref) throws IOException, IllegalStateException;
+
+  class VersionedObject<X> {
+    private X x;
+    private String version;
+
+    public VersionedObject(X x, String version) {
+      this.x = x;
+      this.version = version;
+    }
+
+    public X getX() {
+      return x;
+    }
+
+    public String getVersion() {
+      return version;
+    }
+  }
+
+  @Override
+  public Optional<R> get(NamespaceResourceReference ref)
+      throws AccessException, IOException {
+    Optional<VersionedObject<T>> obj = TransactionRunners.run(transactionRunner, context -> {
+      return getMetadataInternal(context, ref);
+    }, IOException.class);
+    if (!obj.isPresent()) {
+      return Optional.empty();
+    }
+    String secureStoreObjKey = constructSecureStoreKey(ref, obj.get().getVersion());
+    byte[] secureData;
+    try {
+      secureData = secureStore.getData(NamespaceId.SYSTEM.getNamespace(), secureStoreObjKey);
+    } catch (Exception e) {
+      throw new IOException(
+          String.format("Failed to fetch data from secure store for key '%s'",
+              secureStoreObjKey), e);
+    }
+    return Optional.of(constructFromSecureData(obj.get().getX(), new String(secureData,
+        StandardCharsets.UTF_8)));
+  }
+
+  @Override
+  public Collection<NamespaceResourceReference> list(String namespace)
+      throws AccessException, IOException {
+    return TransactionRunners.run(transactionRunner, context -> {
+      return listMetadataInternal(context, namespace);
+    }, IOException.class);
+  }
+
+  @Override
+  public void create(NamespaceResourceReference ref, R obj)
+      throws AccessException, IOException, IllegalArgumentException, IllegalStateException {
+    Util.validateResourceName(ref);
+    String version = createIndex(ref);
+    writeSecureData(ref, obj, version);
+    TransactionRunners.run(transactionRunner, context -> {
```

Comment:
one of the problems with having an abstraction over writing to two different systems is that the operation isn't actually atomic, which can cause all sorts of issues. For example, the secure store write can complete while the DB write fails. The overall create fails, but now there is dirty leftover data so it's more like a half-write. It can't be fully fixed by deleting the secure data on failure, as the delete can fail as well. It's also open to race conditions, if two threads try to create the same resource at the same time.

I would recommend separating out the operations instead of trying to put them together. In other words, the caller needs to create the resource first, and can only update metadata for an existing resource.

Reply:
alternatively if it's possible to just use the secure store instead of secure store plus DB, that would also simplify things.

Reply:
This existing design actually adds an index to make versioned writing possible to avoid data synchronization issues. However, after offline syncing, we decided to put all of the resources directly in a structured table instead of leveraging SecureStore.